src/Components/Server/src/Circuits/CircuitPrerenderer.cs

```diff
@@ -1,42 +1,58 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
+using System.Runtime.ExceptionServices;
 using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Server.Prerendering;
 using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Http.Extensions;
 
 namespace Microsoft.AspNetCore.Components.Server.Circuits
 {
     internal class CircuitPrerenderer : IComponentPrerenderer
     {
+        private static object CircuitHostKey = new object();
+
         private readonly CircuitFactory _circuitFactory;
 
         public CircuitPrerenderer(CircuitFactory circuitFactory)
         {
             _circuitFactory = circuitFactory;
         }
 
-        public async Task<IEnumerable<string>> PrerenderComponentAsync(ComponentPrerenderingContext prerenderingContext)
+        public async Task<ComponentPrerenderResult> PrerenderComponentAsync(ComponentPrerenderingContext prerenderingContext)
         {
             var context = prerenderingContext.Context;
-            var circuitHost = _circuitFactory.CreateCircuitHost(
-                context,
-                client: null,
-                GetFullUri(context.Request),
-                GetFullBaseUri(context.Request));
+            var circuitHost = GetOrCreateCircuitHost(context);
 
             // For right now we just do prerendering and dispose the circuit. In the future we will keep the circuit around and
             // reconnect to it from the ComponentsHub.
-            try
-            {
-                return await circuitHost.PrerenderComponentAsync(
+                var renderResult = await circuitHost.PrerenderComponentAsync(
                     prerenderingContext.ComponentType,
                     prerenderingContext.Parameters);
+
+                return new ComponentPrerenderResult(renderResult);
+        }
+
+        private CircuitHost GetOrCreateCircuitHost(HttpContext context)
+        {
+            if(context.Items.TryGetValue(CircuitHostKey, out var existingHost)) {
+                return (CircuitHost)existingHost;
```

Comment:
We reuse the same circuit host for all prerendered components in a given request. This is equivalent to a non-prerendered scenario with multiple components registered.

Reply:
Makes sense